warp10/src/main/java/io/warp10/script/functions/math/ACOS.java

```diff
@@ -14,45 +14,39 @@
 //   limitations under the License.
 //
 
-package io.warp10.script.functions;
+package io.warp10.script.functions.math;
 
 import io.warp10.script.NamedWarpScriptFunction;
-import io.warp10.script.WarpScriptStackFunction;
 import io.warp10.script.WarpScriptException;
 import io.warp10.script.WarpScriptStack;
+import io.warp10.script.WarpScriptStackFunction;
 
-import java.lang.reflect.Method;
+import java.math.BigDecimal;
 
-public class MATH extends NamedWarpScriptFunction implements WarpScriptStackFunction {
+/**
+ * Check https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html for the documentation of this function.
+ * Last java parameter is on top of the stack.
+ */
+public class ACOS extends NamedWarpScriptFunction implements WarpScriptStackFunction {
 
-  //TODO(tce): LambdaMetafactory could be used to make calls a bit faster, but it is only available on Java 8. See https://stackoverflow.com/q/19557829
-  
-  private Method method;
-  
-  public MATH(String name, String methodName) throws WarpScriptException {
+  public ACOS(String name) {
     super(name);
-    
-    try {
-      this.method = Math.class.getMethod(methodName, double.class);
-    } catch (Exception e) {
-      throw new WarpScriptException(e);
-    }
   }
-  
+
   @Override
   public Object apply(WarpScriptStack stack) throws WarpScriptException {
-    Object o = stack.pop();
+    Object op0 = stack.pop();
 
-    if (!(o instanceof Number)) {
-      throw new WarpScriptException(getName() + " operates on a numeric argument.");
+    if (!(op0 instanceof Number)) {
+      throw new WarpScriptException(getName() + " can only operate on numerical values.");
     }
-    
-    try {
-      stack.push(this.method.invoke(null, ((Number) o).doubleValue()));
-    } catch (Exception e) {
-      throw new WarpScriptException(e);
+
+    if ((op0 instanceof Double || op0 instanceof Long || op0 instanceof Float || op0 instanceof Integer || op0 instanceof BigDecimal)) {
```

Comment:
What Number instances cannot be handled?

Reply:
By definition: BigInteger, Byte, and Short and on top of that, only for Java 8, AtomicInteger, AtomicLong,  DoubleAccumulator, DoubleAdder, LongAccumulator, LongAdder.

Should we allow them too?

I think only AtomicInteger could be on the stack with the current code but others could be in the future or in extensions.